File: android/src/main/java/y2k/joyreactor/WebImageView.java
package y2k.joyreactor;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.ImageView;
import kotlin.Unit;
import y2k.joyreactor.platform.ImageRequest;

/**
 * Created by y2k on 9/26/15.
 */
public class WebImageView extends ImageView {

    private static Image EMPTY = new Image("", 0, 0);

    private Invalidator invalidator = new Invalidator();

    public WebImageView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public void setImage(Image image) {
        invalidator.invalidateStateChanges(image == null ? EMPTY : image);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        invalidator.invalidateStateChanges(null);
    }

    private class Invalidator {

        private Image image;
        private int lastWidth;
        private int lastHeight;

        public void invalidateStateChanges(Image newImage) {
            if (newImage == EMPTY) {
                invalidateToEmpty();
            } else {
                if (newImage == null) invalidateOnChangeSize();
                else invalidateOnImageChanged(newImage);
            }
        }

        private void invalidateToEmpty() {
            if (this.image == EMPTY) return;
            setImageDrawable(null);
            new ImageRequest()
                    .setUrl(null)
                    .to(this, bitmap -> Unit.INSTANCE);
        }

        private void invalidateOnChangeSize() {
            if (image == null) return;
            if (lastWidth == getWidth() && lastHeight == getHeight()) return;

            lastWidth = getWidth();
            lastHeight = getHeight();

            setImageDrawable(null);
            new ImageRequest()
                    .setUrl(image)
                    .setSize(lastWidth, lastHeight)
                    .to(this, bitmap -> {
                        setImageBitmap(bitmap);
                        return Unit.INSTANCE;
                    });
        }

        private void invalidateOnImageChanged(Image newImage) {
            if (image != null && image.equals(newImage)) return;

            image = newImage;

            if (lastWidth == 0 || lastHeight == 0) return;

            lastWidth = getWidth();
            lastHeight = getHeight();

            setImageDrawable(null);
            new ImageRequest()
                    .setUrl(image)
                    .setSize(lastWidth, lastHeight)
                    .to(this, bitmap -> {
                        setImageBitmap(bitmap);
                        return Unit.INSTANCE;
                    });
        }
    }
}

File: android/src/main/kotlin/y2k/joyreactor/WebImageView.kt
package y2k.joyreactor

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.util.AttributeSet
import android.widget.ImageView
import y2k.joyreactor.common.ServiceLocator
import y2k.joyreactor.model.Image
import y2k.joyreactor.services.LinksPool
import y2k.joyreactor.services.ImageService
import kotlin.properties.Delegates

/**
 * Created by y2k on 9/26/15.
 */
class WebImageView(context: Context, attrs: AttributeSet) : ImageView(context, attrs) {

    val imageService = ServiceLocator.resolve<ImageService>()
    var image by Delegates.observable(null as Image?) { prop, old, new -> invalidate() }

    private var state = ImageState()

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (isInEditMode) return

        if (state.image != image || state.width != canvas.width || state.height != canvas.height) {
            state = ImageState(image, canvas.width, canvas.height)

            val url = imageService.makeUrl(state.image, state.width, state.height)
            imageService.to<Bitmap>(LinksPool.default, url, this).subscribe { setImageBitmap(it) }
        }
    }

    private data class ImageState(val image: Image? = null, val width: Int = 0, val height: Int = 0)
}